src/config/middleware/experimental-metrics-csv-middleware.ts

```diff
@@ -0,0 +1,164 @@
+  return [
+    'numActiveRequests',
+    'requestType',
+    'status',
+    'startTime',
+    'endTime',
+    'duration',
+    'requestSize',
+    'responseSize',
+  ];
+}
+
+interface RequestMetrics {
+  numActiveRequests: number;
+  requestType: string;
+  status: number;
+  startTime: number;
+  endTime: number;
+  duration: number;
+  requestSize: number;
+  responseSize: number;
+}
+
+class ExperimentalMetricsCsvMiddlewareRequestHandler
+  implements MiddlewareRequestHandler
+{
+  private readonly logger: MomentoLogger;
+  private readonly csvPath: string;
+  private readonly startTime: number;
+  private requestType: string;
+  private requestSize: number;
+  private responseStatusCode: number;
+  private responseSize: number;
+
+  private receivedResponseBody: boolean;
+  private receivedResponseStatus: boolean;
+
+  constructor(logger: MomentoLogger, csvPath: string) {
+    this.logger = logger;
+    this.csvPath = csvPath;
+    ExperimentalMetricsCsvMiddleware.numActiveRequests++;
+    this.startTime = new Date().getTime();
+
+    this.receivedResponseBody = false;
+    this.receivedResponseStatus = false;
+  }
+
+  onRequestBody(request: Message): Promise<Message> {
+    this.requestSize = request.serializeBinary().length;
+    this.requestType = request.constructor.name;
+    return Promise.resolve(request);
+  }
+
+  onRequestMetadata(metadata: Metadata): Promise<Metadata> {
+    return Promise.resolve(metadata);
+  }
+
+  onResponseBody(response: Message | null): Promise<Message | null> {
+    if (response !== null) {
+      this.responseSize = response.serializeBinary().length;
+    } else {
+      this.responseSize = 0;
+    }
+    this.receivedResponseBody = true;
+    if (this.done()) this.recordMetrics();
+    return Promise.resolve(response);
+  }
+
+  onResponseMetadata(metadata: Metadata): Promise<Metadata> {
+    return Promise.resolve(metadata);
+  }
+
+  onResponseStatus(status: StatusObject): Promise<StatusObject> {
+    this.receivedResponseStatus = true;
+    this.responseStatusCode = status.code;
+    if (this.done()) this.recordMetrics();
+    return Promise.resolve(status);
+  }
+
+  private done(): boolean {
+    return this.receivedResponseBody && this.receivedResponseStatus;
+  }
+
+  private recordMetrics(): void {
+    const endTime = new Date().getTime();
+    const metrics: RequestMetrics = {
+      numActiveRequests: ExperimentalMetricsCsvMiddleware.numActiveRequests,
+      requestType: this.requestType,
+      status: this.responseStatusCode,
+      startTime: this.startTime,
+      endTime: endTime,
+      duration: endTime - this.startTime,
+      requestSize: this.requestSize,
+      responseSize: this.responseSize,
+    };
+
+    const csvRow = [
+      metrics.numActiveRequests,
+      metrics.requestType,
+      metrics.status,
+      metrics.startTime,
+      metrics.endTime,
+      metrics.duration,
+      metrics.requestSize,
+      metrics.responseSize,
+    ].join(',');
```

Comment:
I prefer using a csv library for this to get the quoting and separator issues to spec.

Reply:
but then we would be forcing that csv dependency on all of our users and probably causing some of them issues with transitive dependencies.  in this case since we have really tight control over all of these values, this seems preferable.